src/pybind/mgr/dashboard/frontend/CMakeLists.txt

```diff
@@ -47,6 +47,27 @@ function(add_npm_options)
 endfunction(add_npm_options)
 
 if(WITH_SYSTEM_NPM)
+  # build mgr dashboard frontend requires Angular v15 supports node.js versions: 14.20.x, 16.13.x and 18.10.x
+  # And More information Please See https://angular.io/guide/update-to-version-15#angular-v15-supports-nodejs-versions-1420x-1613x-and-1810x
+  find_program(NODEJS_EXECUTABLE NAMES node nodejs)
+  if(NODEJS_EXECUTABLE)
+    execute_process(COMMAND
+            ${NODEJS_EXECUTABLE} --version
+            OUTPUT_VARIABLE NODEJS_VERSION
+            ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+    string(REPLACE "v" "" NODEJS_VERSION ${NODEJS_VERSION})
+    if("${NODEJS_VERSION}" VERSION_LESS "14.20.0")
+      message(FATAL_ERROR "The node version must greater than version v14.20.0")
+    endif()
+    if(("${NODEJS_VERSION}" VERSION_GREATER_EQUAL "16.0.0") AND ("${NODEJS_VERSION}" VERSION_LESS "16.13.0"))
+      message(FATAL_ERROR "The node version must greater than version v16.13.0")
+    endif()
+    if(("${NODEJS_VERSION}" VERSION_GREATER_EQUAL "18.0.0") AND ("${NODEJS_VERSION}" VERSION_LESS "18.10.0"))
+      message(FATAL_ERROR "The node version must greater than version v18.10.0")
+    endif()
```

Comment:
This is going to be really hard to maintain. I'd just leave `npm install` or `ng build` to scream if the dependencies are not met.

Reply:
Angular has developed to version v17.0.x. And the version v15 is the past version and will basically not change; if there is a chance in the future, it will adapt to Inode.js versions 14.[15-19].x or 16.[ 10-12].x, I will submit a PR to change this version at that time. This is a process of continuous contribution.

Reply:
@TomNewChao in my experience, and specially when dealing with external dependendencies and no security risks involved, it's better to let the external deps complain than trying to be overzealous (this is aligned to the [Pythonic EAFP](https://realpython.com/python-lbyl-vs-eafp/) approach).

BTW, if [this is all because my comment here](https://github.com/ceph/ceph/pull/52617#discussion_r1466251761), I'll gladly take it back :P as this change is valuable enough without the version checking, and SYSTEM_NPM is not the default approach.

---

That said, if we definitely want to enforce a specific version range from the Ceph side, I checked Ceph's CMake files and we're using strict version checking only for:
* Python
* C compiler
* Boost

These are core deps and the checks are mostly done to cope with side effects of specific versions (e.g.: https://tracker.ceph.com/issues/45147).

So IMHO rather than "simply" failing at build stage if those deps are missing/incorrect, we should enforce the desired versions by declaring those in the appropriate places, like:
* `ceph.spec.in` (RHEL/SUSE): `BuildRequires: python3-jinja2 >= 2.10`
* `debian/control` (Debian/Ubuntu): `Build-Depends: ... libthrift-dev (>= 0.13.0),`

That would similar to what I suggested [here](https://github.com/ceph/ceph/pull/52617#discussion_r1466251761), and for [Angular 15](https://angular.io/guide/versions):
```spec
BuildRequires:  node<20.0 and (node >= 14.20 or node>=16.13 or node >= 18.10)
BuildRequires:  npm
```
That said, I checked in major distros and:
| Distro | Default nodejs version | Ok | Comment |
|--------|------------------------|----|---------|
| CentOS 9 | 16.20.2 | ✅ | |
| CentOS 8 | 10.23 | ❌   | Requires [enabling modules](https://docs.centos.org/en-US/8-docs/managing-userspace-components/assembly_introduction-to-modules/) |
| Ubuntu 22.04 | 12.22 | ❌ | Requires [NodeSource PPA](https://github.com/nodesource/distributions) |
| Ubuntu 20.04 | 10.19 | ❌ | Requires [NodeSource PPA](https://github.com/nodesource/distributions) |

So this basically means that for CentOS8 and Ubuntu, we need to enable those extra sources at `install-deps.sh` script (and that doesn't work with `rpmbuild` or `dpkg-buildpackage`, right @tchaikov ?)

---

### Summarizing

This is too complicate given that `WITH_SYSTEM_NPM` is disabled by default (instead, [Python `nodeenv` package](https://pypi.org/project/nodeenv/) is used to specify nodejs versions). So, I'd probably remove any version checking

Reply:
Just like the PR I submitted, two methods are used to support ceph dashboard multi-language:
1. Build the dashboard ui while compiling ceph, but it will strongly depend on the versions of Npm and nodejs to support building the dashboard, so we need to determine the version of nodejs. There is no doubt that this is the right approach. but this is going to be tough to maintain in the long run.
2. Follow the existing implementation logic, support dashboard multi-language when executing make-dist, compress it into the release package, and avoid compiling dashboard ui when building ceph, because the default value of `WITH_SYSTEM_NPM` is false, which is the meaning of nodejs version judgment at this time It's not particularly big, because it is turned off by default during build, unless someone sets `WITH_SYSTEM_NPM` to True, and then it needs to ensure the nodejs version by themself.
So, I agree with you to cancel the nodejs version version checking.